### PR TITLE
Prevent infinity loop teleports (anti fools)

### PR DESCRIPTION
**Pull request by MillhioreBT**
With this there will be no way to cause a server crash using teleports loops.
It sounds very silly but what difference does it make?
I like the idea that every time the server avoids more ways of being blocked by any human stupidity.

This was mentioned in the closed Pull Request: #2993

### CODE REVIEW

File: src/teleport.cpp [hunk 1]
```diff
@@ -21,6 +21,7 @@
 
 #include "teleport.h"
 #include "game.h"
+#include <boost/format.hpp>
```
Comment (DSpeichert) [inline src/teleport.cpp:24]: Not sure if this doesn't require explicitly adding this library to cmake.

File: src/teleport.cpp [hunk 2]
```diff
@@ -70,13 +71,36 @@ void Teleport::addThing(Thing* thing)
...
 {
 	Tile* destTile = g_game.map.getTile(destPos);
 	if (!destTile) {
 		return;
 	}
 
+	// Prevent infinity loop (anti fools)
+	if (checkInfinityLoop(destTile)) {
+		const Position pos = getPosition();
+		std::cout << boost::format("Warning: infinity loop teleport. (x:%1%, y:%2%, z:%3%)\n") % pos.getX() % pos.getY() % pos.getZ();
```
Comment (nekiro) [inline src/teleport.cpp:100]: why that boost::format thing? Just use 
```
std::cout << "Warning: infinity loop teleport. (x: " << pos.x << " y: " << pos.y << " z: " << pos.z << ")" std::endl;
```

File: src/teleport.cpp [hunk 3]
```diff
@@ -70,13 +71,36 @@ void Teleport::addThing(Thing* thing)
...
 }
 
+bool Teleport::checkInfinityLoop(Tile* destTile)
+{
+	if (!destTile) {
+		return false;
+	}
+
+	if (Teleport* teleport = destTile->getTeleportItem()) {
+		Position nextDestPos = teleport->getDestPos();
```
Comment (nekiro) [inline src/teleport.cpp:81]: missing reference
Position& nextDestPos can be a const too

File: src/teleport.cpp [hunk 4]
```diff
@@ -70,13 +71,36 @@ void Teleport::addThing(Thing* thing)
...
 void Teleport::addThing(int32_t, Thing* thing)
 {
 	Tile* destTile = g_game.map.getTile(destPos);
 	if (!destTile) {
 		return;
 	}
 
+	// Prevent infinity loop (anti fools)
+	if (checkInfinityLoop(destTile)) {
+		const Position pos = getPosition();
```
Comment (nekiro) [inline src/teleport.cpp:99]: missing reference Position&

File: src/teleport.cpp (hunk 1 above)
Comment (slavidodo) [reply src/teleport.cpp:24]: boost::format is not required at all!

File: src/teleport.cpp [hunk 5]
```diff
@@ -70,13 +70,36 @@ void Teleport::addThing(Thing* thing)
...
+}
+
 void Teleport::addThing(int32_t, Thing* thing)
 {
 	Tile* destTile = g_game.map.getTile(destPos);
 	if (!destTile) {
 		return;
 	}
 
+	// Prevent infinity loop (anti fools)
```
Comment (nekiro) [inline src/teleport.cpp:96]: I think we could get rid of that (anti fools) just leave it as prevent infinity loop

File: src/teleport.cpp [hunk 6]
```diff
@@ -70,13 +70,36 @@ void Teleport::addThing(Thing* thing)
 	return addThing(0, thing);
 }
 
+bool Teleport::checkInfinityLoop(Tile* destTile)
+{
+	if (!destTile) {
```
Comment (nekiro) [inline src/teleport.cpp:75]: this check is redudant, cause you check that in the method which you call that one.

**Pull request merged**